error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MockitoMockClassReference.java

```diff
@@ -0,0 +1,95 @@
+import static java.util.Objects.requireNonNull;
+import static tech.picnic.errorprone.bugpatterns.util.Documentation.BUG_PATTERNS_BASE_URL;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.Iterables;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.ReturnTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+
+/**
+ * A {@link BugChecker} that flags the use of {@link org.mockito.Mockito#mock(Class)} and {@link
+ * org.mockito.Mockito#spy(Class)} where instead the type could be derived from the variable or
+ * field with an explicit type.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary =
+        "Don't unnecessarily pass the type reference to Mockito's `mock(Class)` and spy(Class)`",
+    link = BUG_PATTERNS_BASE_URL + "MockitoMockClassReference",
+    linkType = CUSTOM,
+    severity = SUGGESTION,
+    tags = SIMPLIFICATION)
+public final class MockitoMockClassReference extends BugChecker
+    implements MethodInvocationTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Matcher<MethodInvocationTree> MOCKITO_MOCK_OR_SPY =
+      allOf(
+          argumentCount(1),
+          argument(0, allOf(isSameType(Class.class.getName()), not(isVariable()))),
+          staticMethod().onClass("org.mockito.Mockito").namedAnyOf("mock", "spy"));
+  // XXX: Replace `var` usage with explicit type instead.
+  private static final Matcher<VariableTree> INCOMPATIBLE_VARIABLE =
+      anyOf(ASTHelpers::hasNoExplicitType, MockitoMockClassReference::hasTypeDifference);
+  private static final Matcher<ReturnTree> INCOMPATIBLE_RETURN =
+      MockitoMockClassReference::hasTypeDifference;
+
+  /** Instantiates a new {@link MockitoMockClassReference} instance. */
+  public MockitoMockClassReference() {}
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (!MOCKITO_MOCK_OR_SPY.matches(tree, state)) {
+      return Description.NO_MATCH;
+    }
+
+    Tree parent = state.getPath().getParentPath().getLeaf();
+    if (parent.getKind() == VARIABLE
+        && INCOMPATIBLE_VARIABLE.matches((VariableTree) parent, state)) {
+      return Description.NO_MATCH;
+    } else if (parent.getKind() == RETURN
+        && INCOMPATIBLE_RETURN.matches((ReturnTree) parent, state)) {
+      return Description.NO_MATCH;
+    }
+
+    return describeMatch(tree, SuggestedFix.delete(Iterables.getOnlyElement(tree.getArguments())));
```

Comment:
The code now assumes a match in case the parent is not a `VARIABLE` or `RETURN` statement. So the following statement is also (incorrectly) flagged:
```java
Objects.hash(mock(Integer.class));
```

Reply:
After some rewriting and trying stuff out I found out we are strictly speaking too many cases. So the example in my previous comment is the `Kind.NEW_CLASS` edge case. We should only rewrite things if we are sure it will be correct, so swapping the default here.

While doing so, I changed the setup to not use `Matcher`s and have multiple `hasTypeDifference` methods. PTAL and let me know what you think.